Formula/microk8s.rb

```diff
@@ -0,0 +1,91 @@
+    url "https://files.pythonhosted.org/packages/4e/ab/5d6bc3b697154018ef196f5b17d958fac3854e2efbc39ea07a284d4a6a9b/click-7.1.1.tar.gz"
+    sha256 "8a18b4ea89d8820c5d0c7da8a64b2c324b4dabb695804dbfea19b9be9d88c0cc"
+  end
+
+  resource "idna" do
+    url "https://files.pythonhosted.org/packages/65/c4/80f97e9c9628f3cac9b98bfca0402ede54e0563b56482e3e6e45c43c4935/idna-2.7.tar.gz"
+    sha256 "684a38a6f903c1d71d6d5fac066b58d7768af4de2b832e426ec79c30daa94a16"
+  end
+
+  resource "jsonschema" do
+    url "https://files.pythonhosted.org/packages/58/0d/c816f5ea5adaf1293a1d81d32e4cdfdaf8496973aa5049786d7fdb14e7e7/jsonschema-2.5.1.tar.gz"
+    sha256 "36673ac378feed3daa5956276a829699056523d7961027911f064b52255ead41"
+  end
+
+  resource "macholib" do
+    url "https://files.pythonhosted.org/packages/0d/fe/61e8f6b569c8273a8f2dd73921738239e03a2acbfc55be09f8793261f269/macholib-1.14.tar.gz"
+    sha256 "0c436bc847e7b1d9bda0560351bf76d7caf930fb585a828d13608839ef42c432"
+  end
+
+  resource "progressbar33" do
+    url "https://files.pythonhosted.org/packages/71/fc/7c8e01f41a6e671d7b11be470eeb3d15339c75ce5559935f3f55890eec6b/progressbar33-2.4.tar.gz"
+    sha256 "51fe0d9b3b4023db2f983eeccdfc8c9846b84db8443b9bee002c7f58f4376eff"
+  end
+
+  resource "requests" do
+    url "https://files.pythonhosted.org/packages/97/10/92d25b93e9c266c94b76a5548f020f3f1dd0eb40649cb1993532c0af8f4c/requests-2.20.0.tar.gz"
+    sha256 "99dcfdaaeb17caf6e526f32b6a7b780461512ab3f1d992187801694cba42770c"
+  end
+
+  resource "requests-unixsocket" do
+    url "https://files.pythonhosted.org/packages/f3/94/67d781fb32afbee0fffa0ad9e16ad0491f1a9c303e14790ae4e18f11be19/requests-unixsocket-0.1.5.tar.gz"
+    sha256 "a91bc0138f61fb3396de6358fa81e2cd069a150ade5111f869df01d8bc9d294c"
+  end
+
+  resource "simplejson" do
+    url "https://files.pythonhosted.org/packages/f0/07/26b519e6ebb03c2a74989f7571e6ae6b82e9d7d81b8de6fcdbfc643c7b58/simplejson-3.8.2.tar.gz"
+    sha256 "d58439c548433adcda98e695be53e526ba940a4b9c44fb9a05d92cd495cdd47f"
+  end
+
+  resource "toml" do
+    url "https://files.pythonhosted.org/packages/b9/19/5cbd78eac8b1783671c40e34bb0fa83133a06d340a38b55c645076d40094/toml-0.10.0.tar.gz"
+    sha256 "229f81c57791a41d65e399fc06bf0848bab550a9dfd5ed66df18ce5f05e73d5c"
+  end
+
+  resource "urllib3" do
+    url "https://files.pythonhosted.org/packages/8a/3c/1bb7ef6c435dea026f06ed9f3ba16aa93f9f4f5d3857a51a35dfa00882f1/urllib3-1.24.3.tar.gz"
+    sha256 "2393a695cd12afedd0dcb26fe5d50d0cf248e5a66f75dbd89a3d4eb333a61af4"
+  end
+
+  resource "wheel" do
+    url "https://files.pythonhosted.org/packages/75/28/521c6dc7fef23a68368efefdcd682f5b3d1d58c2b90b06dc1d0b805b51ae/wheel-0.34.2.tar.gz"
+    sha256 "8788e9155fe14f54164c1b9eb0a319d98ef02c160725587ad60f14ddc57b6f96"
+  end
+
+  def install
+    venv = virtualenv_create(libexec, "python3")
+    venv.pip_install resources
+    venv.pip_install_and_link buildpath/"installer"
+  end
+
+  test do
+    system "#{bin}/microk8s", "uninstall"
```

Comment:
Can you check this has any effect at all?

Reply:
No we can't.  Nor install as it requires a virtualisation to be working, which it won't be on most test runners.

Reply:
Will it fail in a predictable way?

Reply:
All we can test is that the binary exits 0

Reply:
It exits with a succes status if it fails?

Reply:
No.

Reply:
In that case I don't know what the problem is. Just do something that fails and check the output

Reply:
To reiterate I can't test anything except `microk8s uninstall`, which does nothing except ensure brew installed the app correctly.  Testing `microk8s install`: 

```Error: Operation not permitted @ dir_s_mkdir - /usr/local/Caskroom/multipass```

Reply:
I've tried asserting for this, but it doesn't work and is only testing brew's confinement, not the microk8s installer.

Reply:
@SMillerDev @dawidd6,  At this point, I have to ask whether or not this is a blocker for homebrew?  We have a bunch of docs and release articles on hold for this and I'm not sure we're getting anywhere.  If this is not acceptable to you guys, we need to know ASAP so we can move to a tap and update the aforementioned.

Thanks

Reply:
If it requires some other software that can't build from source it's not acceptable in homebrew-core. I initially thought multipass was additional software but the current situation means it's always a broken install at first.

Reply:
@SMillerDev that's not entirely true as the actual thing you're installing via brew, works.  When you first runs it, it then gets multipass.  On top of this, we tried to get a cask in a while ago and it was rejected https://github.com/Homebrew/homebrew-cask/pull/76876

So what we're being told is we're not allowed anywhere except our own tap?

Reply:
@SMillerDev Important to add that there is also precedence for this exact situation.  `snap craft.rb` literally does the same as this package.

Reply:
It's incomplete software that does some more package management on the side to make it complete. I'm qualifying that as: https://docs.brew.sh/Acceptable-Formulae#we-dont-like-install-scripts-that-download-unversioned-things

I think the best way forward is to get multipass in homebrew-core and depend on that. Since I don't see anything there that requires it being a cask.

Reply:
If you don't have time for that your own tap is the best way forward.

Reply:
@SMillerDev there are proprietary components.

Reply:
@SMillerDev could you please confirm that this is being allowed in certain packages but not others?

Reply:
It's not allowed in any packages. If this is how snapcraft works it should be removed.

Reply:
@SMillerDev To clarify, neither "downloads unversioned things".  Both are wrappers that require multipass and are installed via brew cask if not present on the machine.

Reply:
I get that it doesn't exactly match the rule but the point remains that the `brew install` command should be the entire install. If your tool needs something else at runtime that should have been declared as a dependency. (and you can't depend on anything outside homebrew-core in homebrew-core)

Reply:
> @SMillerDev there are proprietary components.

Of what: microk8s or multipass? It's a bit dubious for inclusion as open source software (which homebrew/core must be) if it requires closed source software to work.

Reply:
@MikeMcQuaid in multipass, which is why it's a cask and not a formula.

Reply:
I think if this formula cannot function without proprietary software it is not a good fit for homebrew/core. I've opened a PR to adjust the documentation accordingly: https://github.com/Homebrew/brew/pull/7195

Reply:
@MikeMcQuaid but that leaves at least 2 formulas, including this one, and I would bet a fair few more, homeless.  We're not allowed to make casks of these (see PR attached earlier in this thread), nor are we allowed to make formulas.

Reply:
@joedburg You can make an external tap or you can adjust your open source application so it does not depend on proprietary software (which I have no idea how possible that is).

Reply:
@MikeMcQuaid it's not possible at all.  as I mention, it's not a difficult argument to make that most packages in Formula are, at somewhere in the stack, reliant on closed source software.  it's quite frustrating that we're the guinea pig on this as we believe this formula ticks the correct boxes (until those boxes were changed) and meets precedence set by other formulas.

Reply:
It sucks that formulae have to be reviewed by humans that make mistakes. None of the existing formulae should be seen as precedence though, everything is reviewed along the rules that are in place at the time. Just because nobody spotted that snapcraft downloads software at runtime doesn't mean that it's allowed.

Reply:
@SMillerDev except, in this case, the rules have literally been changed beneath us.

Reply:
> it's not possible at all

It may not be possible due to organisational restrictions but I find it hard to believe if this tool had to be built from the ground up to only rely on open source software that that would be impossible.

> it's quite frustrating that we're the guinea pig on this as we believe this formula ticks the correct boxes (until those boxes were changed)
> the rules have literally been changed beneath us.

See https://docs.brew.sh/Acceptable-Formulae#sometimes-there-are-exceptions which makes it clear that these are not a concrete list of "rules", that there are exceptions (in both directions) and that ultimately what maintainers decide trumps documentation.

Reply:
> It may not be possible due to organisational restrictions but I find it hard to believe if this tool had to be built from the ground up to only rely on open source software that that would be impossible.

PRs are always welcome.